test/bootstrap_form_group_test.rb

```diff
@@ -147,6 +147,32 @@ class BootstrapFormGroupTest < ActionView::TestCase
     assert_equivalent_xml expected, @builder.text_field(:email, prepend: '$', append: '.00')
   end
 
+  test "adding both prepend and append text with validation error" do
+    @user.email = nil
+    assert @user.invalid?
+
+    expected = <<-HTML.strip_heredoc
+      <form accept-charset="UTF-8" action="/users" class="new_user" id="new_user" method="post" role="form">
+        <input name="utf8" type="hidden" value="&#x2713;"/>
+        <div class="form-group">
+          <label class="required" for="user_email">Email</label>
+          <div class="input-group">
+            <div class="input-group-prepend">
+              <span class="input-group-text">$</div>
+            </div>
+            <input class="form-control is-invalid" id="user_email" name="user[email]" type="text" />
+            <div class="input-group-append">
+              <span class="input-group-text">.00</span>
+            </div>
+            <div class="invalid-feedback">can't be blank, is too short (minimum is 5 characters)</span>
+          </div>
+        </div>
+      </form>
+    HTML
+    # TODO: We should build the @builder properly from `bootstrap_form_for`, so it's easier to test errors.
+    assert_equivalent_xml expected, bootstrap_form_for(@user) { |f| f.text_field :email, prepend: '$', append: '.00' }
```

Comment:
I would render `@builder.text_field :email, prepend: "foo", append: "bar"` instead of doing it though form_helper. Less html that way. All tests should just call helpers directly.

Reply:
That's how I started, but then it renders the default Rails error markup. `@builder` is missing the `bootstrap_form` override of the default Rails error rendering. I was going to enter an issue to change the way we make the builder, so it would render properly without the extra HTML, which I agree is a unnecessary and can be a pain.

Reply:
You can put code used for error suppression (https://github.com/bootstrap-ruby/bootstrap_form/blob/master/lib/bootstrap_form/helper.rb#L45) into `setup` and `teardown` for tests that need it. You'll want to extract all tests that deal with error rendering into a separate file though.

Reply:
That's a good suggestion. In the spirit of limiting the scope of each PR, I continued to just tolerate a few extra lines of HTML in each test, but we should revisit the suggestion for easier error testing in another PR.